### ISSUE TITLE
Generated models and value semantics

### ISSUE THREAD

**Issue by cuva** [apollographql/apollo-ios#87]
I was wondering about the possibility of changing the way properties are generated in the `Data` model. At the moment they are all defined as constant (`let`) - which makes a lof of sense from a consumer point of view: data is parsed, serialized and should be constant past this point - but doesn't offer much flexibility.

However using `var` for `Data` properties, the schema definition could still enforce that its returned data is `constant` without necessarily restricting it on the client for later operations.

For example:

```swift
class MyQuery {
  struct Data {
    var myField: String
  }
}

public struct GraphQLResult<Data> {
  /// The typed result data, or `nil` if an error was encountered that prevented a valid response.
  public let data: Data?
  [...]
}
```

In this case, `GraphQLResult.data` will still be constant - as will any of its sub-properties - but one could have the possibility to mutate such value later on if this use case were to be necessary by explicitly copying it into a variable.

The current use case I'm trying to solve doesn't seem far stretched: we're using the GraphQL models as provided by the API (i.e dumb client) and want to allow optimistic mutation of the UI (i.e changing the local model when the user performs and action, such as liking a post - thus changing the current model to what we expect it to become - and posting the request after and managing its response). We're currently using a variable proxy object to perform these changes but it ends up adding a lot of complexity and boilerplate that could potentially be avoided.

Alternatively, if this behavior shouldn't be considered as _normal_  - this could be defined as an option to `apollo-codegen`.

Thoughts?

**Comment by martijnwalraven** [comment]
Yep, this is exactly what I've been working on! Hoping to finally open a PR with the changes today. 

It uses computed properties and a private `snapshot` property however, to make sure we can also modify objects through fragments.

The PR also includes an [imperative store API](https://dev-blog.apollodata.com/apollo-clients-new-imperative-store-api-6cb69318a1e3) similar to the one recently added to the JavaScript client, so you can use the generated models to manually make changes to the store.

**Comment by cuva** [comment]
Awesome - thanks @martijnwalraven! Looking forward to it 💯

**Comment by martijnwalraven** [comment]
@cuva: Sorry it has taken a little longer to open a PR, but WIP is finally here #89!

**Comment by martijnwalraven** [comment]
I'm going to close this issue, but we just released Apollo iOS 0.6.0-beta.1, which includes these changes. Make sure to use `npm install -g apollo-codegen@next` until it is out of beta.
